### PR TITLE
Port `#[derive(FromRequest)]`

### PR DESCRIPTION
**Pull request by davidpdrsn**
This ports `#[derive(FromRequest)]` to the new traits.

It always implements `FromRequest` (we might want `#[derive(FromRequestParts)]` but that can be done separately) and only allows the last field to implement `FromRequest`. I.e.:

```rust
// doesn't work
// #[derive(FromRequest)]
// struct Foo {
//     body: String,
//     ext: Extension<bool>,
// }

// does work
#[derive(FromRequest)]
struct Foo {
    ext: Extension<bool>,
    body: String,
}
```

This probably isn't the most intuitive thing ever but the compile error is good so should be pretty clear. Its also documented.

I also ran into some issues with generating custom rejections. Specifically you do something like this

```rust
#[derive(FromRequest)]
struct Foo {
    ext: Extension<bool>,
    body: String,
}
```

We'd generate a rejection like this

```rust
enum FooRejection {
    Ext(<Extension<bool> as FromRequest<(), Body>>::Rejection),
    Body(<String as FromRequest<(), Body>>::Rejection),
}
```

However `<Extension<bool> as FromRequest<(), Body>>::Rejection` doesn't work because like we've seen before `Extension` doesn't implement `FromRequest<(), Body, /* Once */>`, it implements `FromRequest<(), Body, Mut>`.

Not sure there is a solution to this so I just decided to not generate custom rejections and instead use `Rejection = Response`. If users want their own rejection type they can use

```rust
#[derive(FromRequest)]
#[from_request(rejection(MyRejection))]
struct Foo {
    ext: Extension<bool>,
    body: String,
}

enum MyRejection { ... }
```

That also simplifies our code a ton because we don't need all the `#[from_request(rejection_derive(!Display, !Error)]` stuff.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
And with this `cargo t --all-features -p axum-macros` now works!

**Review by jplatte — APPROVED** [review]
Heh, when working on the enum impls I was wondering whether they should have an `EitherN` rejection type. But it's probably good to not have that until there is a clear use case, which I honestly don't expect to ever come, considering you should be able to extract things individually in any situation where you would observe the rejection type), and the same applies here… Probably even to a stronger degree because you can use `#[from_request(rejection(Foo))]` and a separate derive macro to generate the `From` implementations.

That's all to say: I fully agree with using `Rejection = Response`!

File: axum-macros/src/from_request/attr.rs [hunk 1]
```diff
@@ -16,13 +16,11 @@ pub(crate) enum FromRequestContainerAttr {
         rejection: Option<syn::Path>,
     },
     Rejection(syn::Path),
-    RejectionDerive(kw::rejection_derive, RejectionDeriveOptOuts),
     None,
 }
 
 pub(crate) mod kw {
     syn::custom_keyword!(via);
-    syn::custom_keyword!(rejection_derive);
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum-macros/src/from_request/attr.rs:25]: Think I'll rather just remove it and describe in the changelog how to update.

[the remark this answers is not in the document]
Comment (jplatte) [reply axum-macros/src/from_request/attr.rs:25]: Alright

**Pull request merged**